dbms/src/Dictionaries/DictionaryBlockInputStream.h

```diff
@@ -0,0 +1,290 @@
+    return ColumnPtr(std::move(column_vector));
+}
+
+template <class DictionaryType, class Key>
+template <class Getter>
+ColumnPtr DictionaryBlockInputStream<DictionaryType, Key>::getColumnFromStringAttribute(
+    Getter getter, const PaddedPODArray<Key> & ids,
+    const ConstColumnPlainPtrs & keys, const DataTypes & data_types,
+    const DictionaryAttribute& attribute, const DictionaryType& dictionary)
+{
+    auto column_string = std::make_unique<ColumnString>();
+    auto ptr = column_string.get();
+    callGetter(getter, ids, keys, data_types, ptr, attribute, dictionary);
+    return ColumnPtr(std::move(column_string));
+}
+
+template <class DictionaryType, class Key>
+ColumnPtr DictionaryBlockInputStream<DictionaryType, Key>::getColumnFromIds(const PaddedPODArray<Key>& ids)
+{
+    auto column_vector = std::make_unique<ColumnVector<UInt64>>();
+    column_vector->getData().reserve(ids.size());
+    for (UInt64 id : ids) 
+    {
+        column_vector->insert(id);
+    }
+    return ColumnPtr(std::move(column_vector));
+}
+
+template <class DictionaryType, class Key>
+void DictionaryBlockInputStream<DictionaryType, Key>::fillKeyColumns(
+    const std::vector<StringRef> & keys, const DictionaryStructure& dictionary_structure,
+    ColumnsWithTypeAndName & columns)
+{
+    for (const DictionaryAttribute & attribute: *dictionary_structure.key)
+    {
+        #define ADD_COLUMN(TYPE) columns.push_back( \
+            ColumnWithTypeAndName(std::move(std::make_unique<ColumnVector<TYPE>>()), attribute.type, attribute.name))
+        switch (attribute.underlying_type)
+        {
+            case AttributeUnderlyingType::UInt8: ADD_COLUMN(UInt8); break;
+            case AttributeUnderlyingType::UInt16:ADD_COLUMN(UInt16); break;
+            case AttributeUnderlyingType::UInt32: ADD_COLUMN(UInt32); break;
+            case AttributeUnderlyingType::UInt64: ADD_COLUMN(UInt64); break;
+            case AttributeUnderlyingType::Int8: ADD_COLUMN(Int8); break;
+            case AttributeUnderlyingType::Int16: ADD_COLUMN(Int16); break;
+            case AttributeUnderlyingType::Int32: ADD_COLUMN(Int32); break;
+            case AttributeUnderlyingType::Int64: ADD_COLUMN(Int64); break;
+            case AttributeUnderlyingType::Float32: ADD_COLUMN(Float32); break;
+            case AttributeUnderlyingType::Float64: ADD_COLUMN(Float64); break;
+            case AttributeUnderlyingType::String: 
+            {
+                columns.push_back(ColumnWithTypeAndName(
+                    std::move(std::make_unique<ColumnString>()), attribute.type, attribute.name));
+                break;
+            }
+        }
+    }
+
+    for (const auto & key : keys)
+    {
+        auto ptr = key.data;
+        for (const auto & column : columns)
+            ptr = column.column->deserializeAndInsertFromArena(ptr);
+        if (ptr != key.data + key.size)
+            throw Exception("DictionaryBlockInputStream: unable to deserialize data");
```

Comment:
No error code.
Better to totally remove this check, because it is in inner loop.